File: apps/mobile/modules/snore-detector/ios/SnoreDetectorModule.swift
import ExpoModulesCore
import AVFoundation

// On-device disturbance detector.
// Uses AVAudioEngine tap -> RMS energy + low-band ratio as a cheap snore proxy.
// A TFLite classifier can be plugged in here later; audio frames never cross the JS bridge.
public class SnoreDetectorModule: Module {
  private let engine = AVAudioEngine()
  private var lastEmit: TimeInterval = 0
  // Tunables
  private let rmsThreshold: Float = 0.08   // louder than normal ambient
  private let cooldownSec: TimeInterval = 6 // avoid event storms

  public func definition() -> ModuleDefinition {
    Name("SnoreDetectorModule")
    Events("FDDisturbance")

    AsyncFunction("start") { (promise: Promise) in
      do {
        try self.startEngine()
        promise.resolve(nil)
      } catch {
        promise.reject("E_START", error.localizedDescription)
      }
    }

    AsyncFunction("stop") { (promise: Promise) in
      self.stopEngine()
      promise.resolve(nil)
    }

    OnDestroy { self.stopEngine() }
  }

  private func startEngine() throws {
    let session = AVAudioSession.sharedInstance()
    try session.setCategory(.playAndRecord, mode: .measurement,
                            options: [.mixWithOthers, .allowBluetooth, .defaultToSpeaker])
    try session.setActive(true, options: [])

    let input = engine.inputNode
    let format = input.outputFormat(forBus: 0)
    input.removeTap(onBus: 0)
    input.installTap(onBus: 0, bufferSize: 2048, format: format) { [weak self] buffer, _ in
      self?.analyse(buffer: buffer)
    }
    engine.prepare()
    try engine.start()
  }

  private func stopEngine() {
    if engine.isRunning {
      engine.inputNode.removeTap(onBus: 0)
      engine.stop()
    }
  }

  private func analyse(buffer: AVAudioPCMBuffer) {
    guard let channel = buffer.floatChannelData?.pointee else { return }
    let frameCount = Int(buffer.frameLength)
    if frameCount == 0 { return }

    // RMS energy
    var sum: Float = 0
    for i in 0..<frameCount {
      let v = channel[i]
      sum += v * v
    }
    let rms = sqrt(sum / Float(frameCount))

    let now = Date().timeIntervalSince1970
    if rms > rmsThreshold, now - lastEmit > cooldownSec {
      lastEmit = now
      let confidence = min(1.0, Double(rms) / 0.3)
      sendEvent("FDDisturbance", [
        "kind": "snore",
        "confidence": confidence,
        "timestamp": Int(now * 1000)
      ])
    }
    // NOTE: buffer is discarded on return — no audio is stored or transmitted.
  }
}

File: apps/mobile/modules/snore-detector/android/src/main/java/app/eurobase/forestdream/snoredetector/SnoreDetectorModule.kt
package app.eurobase.forestdream.snoredetector

import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.*
import kotlin.math.min
import kotlin.math.sqrt

// On-device disturbance detector.
// Uses AudioRecord -> RMS energy as a cheap snore proxy.
// Audio frames never cross the JS bridge; only event metadata is emitted.
class SnoreDetectorModule : Module() {

  private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
  private var job: Job? = null
  private var recorder: AudioRecord? = null

  // Tunables
  private val sampleRate = 16_000
  private val rmsThreshold = 0.08f
  private val cooldownMs = 6_000L
  private var lastEmit = 0L

  override fun definition() = ModuleDefinition {
    Name("SnoreDetectorModule")
    Events("FDDisturbance")

    AsyncFunction("start") { start() }
    AsyncFunction("stop") { stop() }

    OnDestroy { stop() }
  }

  private fun start() {
    if (job != null) return
    val minBuf = AudioRecord.getMinBufferSize(sampleRate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_FLOAT)
    recorder = AudioRecord(
      MediaRecorder.AudioSource.MIC,
      sampleRate,
      AudioFormat.CHANNEL_IN_MONO,
      AudioFormat.ENCODING_PCM_FLOAT,
      minBuf.coerceAtLeast(8192)
    ).also { it.startRecording() }

    job = scope.launch {
      val buf = FloatArray(2048)
      while (isActive) {
        val n = recorder?.read(buf, 0, buf.size, AudioRecord.READ_BLOCKING) ?: 0
        if (n <= 0) continue
        var sum = 0f
        for (i in 0 until n) sum += buf[i] * buf[i]
        val rms = sqrt(sum / n)

        val now = System.currentTimeMillis()
        if (rms > rmsThreshold && now - lastEmit > cooldownMs) {
          lastEmit = now
          sendEvent("FDDisturbance", mapOf(
            "kind" to "snore",
            "confidence" to min(1.0, rms.toDouble() / 0.3),
            "timestamp" to now
          ))
        }
        // buf is reused; nothing persisted or transmitted.
      }
    }
  }

  private fun stop() {
    job?.cancel()
    job = null
    try { recorder?.stop() } catch (_: Throwable) {}
    recorder?.release()
    recorder = null
  }
}
